controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java

```diff
@@ -614,49 +614,48 @@ private Scope getScope(final String scopeName) {
-                                        new AbstractMap.SimpleEntry<>(
-                                                segment.getKeyStart(),
-                                                segment.getKeyEnd())))
-                        .collect(Collectors.toList()));
-    }
-
-    private CompletableFuture<Long> findNumSplits(String scopeName, String streamName, Executor executor) {
+    private CompletableFuture<Pair<Long, Long>> findNumSplitsMerges(String scopeName, String streamName, Executor executor) {
         return getScaleMetadata(scopeName, streamName, null, executor).thenApply(scaleMetadataList -> {
-            long size = scaleMetadataList.size();
+            int size = scaleMetadataList.size();
             long totalNumSplits = 0;
-
-            for (int i = 0, j = 1; j < size; j++) {
-                long countSplitsEventI = scaleMetadataList.get(i).getSegments().size();
-                long countSplitsEventJ = scaleMetadataList.get(j).getSegments().size();
-                totalNumSplits = (countSplitsEventI < countSplitsEventJ) ? (countSplitsEventJ - countSplitsEventI) : 0;
+            long totalNumMerges = 0;
+            List<Segment> segmentList1;
+            List<Segment> segmentList2;
+            boolean isDescendingOrder = (size > 1) ?
+                    (scaleMetadataList.get(0).getTimestamp() < scaleMetadataList.get(1).getTimestamp()) : true;
+
+            for (int i = 0; i < size - 1; i++) {
+                segmentList1 = scaleMetadataList.get(i).getSegments();
+                segmentList2 = scaleMetadataList.get(i+1).getSegments();
+                if (isDescendingOrder) {
+                    totalNumSplits += findSegmentSplitsMerges(segmentList2, segmentList1);
+                    totalNumMerges += findSegmentSplitsMerges(segmentList1, segmentList2);
+                } else {
+                    totalNumSplits += findSegmentSplitsMerges(segmentList1, segmentList2);
+                    totalNumMerges += findSegmentSplitsMerges(segmentList2, segmentList1);
+                }
             }
 
-            return totalNumSplits;
+            return new ImmutablePair<>(totalNumSplits, totalNumMerges);
         });
     }
 
-    private CompletableFuture<Long> findNumMerges(String scopeName, String streamName, Executor executor) {
-        return getScaleMetadata(scopeName, streamName, null, executor).thenApply(scaleMetadataList -> {
-            long size = scaleMetadataList.size();
-            long totalNumMerges = 0;
-
-            for (int i = 0, j = 1; j < size; j++) {
-                long countMergesEventI = scaleMetadataList.get(i).getSegments().size();
-                long countMergesEventJ = scaleMetadataList.get(j).getSegments().size();
-                totalNumMerges = (countMergesEventI > countMergesEventJ) ? (countMergesEventI - countMergesEventJ) : 0;
+    private int findSegmentSplitsMerges(List<Segment> segmentsList1, List<Segment> segmentsList2) {
+        int count = 0;
```

Comment:
If i understand your example you are saying 
`i` has S0 
`i+1` has S, S'
`i+2` has S2 which is a merge of S and S'

the algo will take s0 and find that it has 2 overlaps in `i+1`. So there is a split. 
for computing merge, it will reverse and compute splits between i + 1 and i.. Since both segments S and S' will have exactly one overlap, the split count will be 0 for either. So number of merges = 0.
now between `i +1` and `i+2` ->
splits = 0 (because S has 1 overlap and S' has 1 overlap)
merges = 1 (because S2 has 2 overlaps - S and S').